appmap/_implementation/generation.py

```diff
@@ -0,0 +1,117 @@
+"""Generate an AppMap"""
+import json
+from collections.abc import MutableMapping
+
+from . import metadata
+
+from .event import Event, serialize_event
+
+
+class ClassMapSet(MutableMapping):
+    def __init__(self):
+        self.dict = dict()
+
+    def __delitem__(self, k):
+        del self[k]
+
+    def __getitem__(self, k):
+        return self.dict[k]
+
+    def __setitem__(self, k, v):
+        self.dict[k] = v
+
+    def __iter__(self):
+        return self.dict.__iter__()
+
+    def __len__(self):
+        return len(self.dict)
+
+
+class ClassMapEntry:
+    def __init__(self, key, name, entry_type):
+        self.key = key
+        self.name = name
+        self.type = entry_type
+
+    def __eq__(self, other):
+        return self.key == other.key
+
+    def __hash__(self):
+        return hash(self.key)
+
+
+class PackageEntry(ClassMapEntry):
+    def __init__(self, name):
+        super().__init__(name, name, 'package')
+        self.children = ClassMapSet()
+
+    __hash__ = ClassMapEntry.__hash__
```

Comment:
@apotterri won't `__hash__` be inherited?

Reply:
No, per https://docs.python.org/3/reference/datamodel.html#object.__hash__ , which says:
> If a class that overrides __eq__() needs to retain the implementation of __hash__() from a parent class, the interpreter must be told this explicitly by setting __hash__ = <ParentClass>.__hash__.

Reply:
from what I have understood, if the subclass overrides `__eq__`, only then its `__hash__` is set to None.

```>>> class ClassMapEntry:
...     def __init__(self, key, name, entry_type):
...         self.key = key
...         self.name = name
...         self.type = entry_type
...     def __eq__(self, other):
...         return self.key == other.key
...     def __hash__(self):
...         return hash(self.key)
...
>>> class PackageEntry(ClassMapEntry):
...     def __init__(self, name):
...         super().__init__(name, name, 'package')
...         self.children = ClassMapSet()
...
>>> PackageEntry.__hash__
<function ClassMapEntry.__hash__ at 0x10acc6e50>
>>> class ClassEntry(ClassMapEntry):
...     def __init__(self, name):
...         super().__init__(name, name, 'class')
...         self.children = ClassMapSet()
...     def __eq__(self, other):
...         pass
...
>>> ClassEntry.__hash__
>>> PackageEntry.__hash__
<function ClassMapEntry.__hash__ at 0x10acc6e50>```

Reply:
anyways, doesn't hurt either ways

Reply:
I confirmed that you're correct, subclasses of `ClassMapEntry` use its `__hash__` method. I've taken the assignment out of the subclasses.

I've taken your other suggestion, though, and switched to dataclasses, eliminating the need for any of this.